core/src/setup/Kokkos_Setup_HIP.hpp

```diff
@@ -27,6 +27,8 @@
 #define KOKKOS_LAMBDA [=] __host__ __device__
 #define KOKKOS_CLASS_LAMBDA [ =, *this ] __host__ __device__
 
+#define KOKKOS_DEDUCTION_GUIDE __host__ __device__
```

Comment:
I understand your point, just wondering if annotating something that is not a function by `KOKKOS_FUNCTION` would be confusing ?

Reply:
It just follows the semantics for deduction guides so it's as confusing and more consistent.

Reply:
Also, we should limit the number of (public) macros we define since they are a pain to deal with later if we consider removing them since we can't deprecate them.

Reply:
> , using KOKKOS_FUNCTION instead of introducing a new macro matches more closely the intent of deduction guides (by replicating the constructor's signature).

I agree with @masterleinad on this. 

I don't really see how introducing a new macro improves the code. Worse it makes a bug more likely. Both this PR and mdspan fix the issue for HIP but not for CUDA clang...

Reply:
Indeed i forgot the case of CUDA clang.

Is it safe though to assume that compilers need the same annotations between deduction guides and constructors ?

Reply:
> Is it safe though to assume that compilers need the same annotations between deduction guides and constructors ?

That's what we are observing, right?

Reply:
So far yes i agree

Reply:
> It just follows the semantics for deduction guides

I don't understand that argument.  Did you mean **syntax**?  Please clarify.

To the consistency argument, I would like to point out there is prior art in Kokkos, with `KOKKOS_DEFAULTED_FUNCTION` and `KOKKOS_INLINE_DELETED_FUNCTION`
<details>
<summary>(Blame game for naming that `KOKKOS_INLINE_DELETED_FUNCTION` macro)</summary>
https://github.com/kokkos/kokkos/commit/3a439317142573f0f6b39e2cd61934fb5e56ca72
</details>

You could argue that these are not needed any more and we are stuck with them, but I also want to look back and see that these were also special syntaxes for the compiler for which it was also not clear what to do about `__host__ __device__` annotations (especially for deleted functions!).  Experience has shown that compilers behavior changed and we had to discriminate workarounds based on vendor toolchain versions.

If we use `KOKKOS_FUNCTION` to decorate our deduction guides, teach our users to do so, and there is even only one compiler that we need to support and that does not like the host device annotation, then we will have to do the distinction.  We will be able to fix it in our codebase but it will be harder to propagate downstream.  
That said I have no evidence it will break and I haven't surveyed what currently works.

> Worse it makes a bug more likely. Both this PR and mdspan fix the issue for HIP but not for CUDA clang...

I don't see how it makes it more bug prone.  Please elaborate.
Regarding Clang+CUDA, did you check that we see the issue when the annotation are missing.  I would be surprised if that was not considered in the reference implementation of `mdspan`.  That is possible but I expect someone would have asked the question.  @crtrott do you remember?

I don't like the proliferation of public macros either but I tend to think it is justified in this case.

Reply:
> I don't see how it makes it more bug prone. Please elaborate.

Different people wrote different versions of this PR and they both forgot to fix CUDA clang. If you didn't introduce the new macro and instead use `KOKKOS_FUNCTION`, the PR would have fixed the issue.

> Regarding Clang+CUDA, did you check that we see the issue when the annotation are missing

If you look at the godbolt code that @tpadioleau posted. You can see that he is using CUDA clang to reproduce the error not HIP.

Reply:
> > Regarding Clang+CUDA, did you check that we see the issue when the annotation are missing
> 
> If you look at the godbolt code that @tpadioleau posted. You can see that he is using CUDA clang to reproduce the error not HIP.

Haha damn.  Missed the compiler explorer link, I was on my cell.